src/cgutils.cpp

```diff
@@ -853,6 +853,31 @@ static Value *emit_bounds_check(const jl_cgval_t &ainfo, jl_value_t *ty, Value *
     return im1;
 }
 
+static void emit_inexacterror_unless(Value *cond, jl_value_t *type, jl_value_t *x, jl_codectx_t *ctx)
+{
+    BasicBlock *failBB = BasicBlock::Create(jl_LLVMContext,"fail",ctx->f);
+    BasicBlock *passBB = BasicBlock::Create(jl_LLVMContext,"pass");
+    builder.CreateCondBr(cond, passBB, failBB);
+    builder.SetInsertPoint(failBB);
+    Value *fname_val = stringConstPtr(ctx->funcName);
+#if JL_LLVM_VERSION >= 30700
+    builder.CreateCall(prepare_call(jlinexacterror_func),
+                       { fname_val, literal_pointer_val(type),
+                         boxed(x, ctx, false)});
+#else
+    builder.CreateCall3(prepare_call(jlinexacterror_func),
+                        fname_val, literal_pointer_val(type),
+                        boxed(x, ctx, false));
```

Comment:
The `jl_value_t*` is a julia object, in codegen they generally represent the expression so they should not be used to construct the error. You need to construct the error using a julia value at runtime, which is created by evaluating the expression represented by the `jl_value_t*` at runtime, which will be represent as a `Value*`. The `Value*` should also be a boxed `jl_value_t*` rather than a unboxed one. This information is kept in the `jl_cgval_t`, which is a wrapper around `Value*` that also includes information about which type of representation it holds.

For the `generic_trunc` one, you can get the `jl_cgval_t` if you inline the `auto_unbox(x)`. For the `emit_untyped_intrinsic` case, you can construct the `jl_cgval_t` from the `Value *x` and the `jl_value_t *xtyp`, however, it seems better to make the caller to pass in the `jl_cgval_t` directly since it's always available.

From `jl_cgval_t`, `boxed` should give you the right pointer to be passed to the C function that constructs the exception.